lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -996,146 +1010,66 @@ SymbolicValue ConstExprFunctionState::getConstantValue(SILValue value) {
+///   * the stores' value(s) can be const-evaluated;
+/// Then: initializes the memory at `addr` and returns {true, None}.
 ///
-///  Note unlike getConstAddrAndLoadResult(), this method does *not*
-///  const-evaluate the input `addr` by evaluating its operand first, such as %7
-///  above. Instead, it finds a user of %8 who is the initializer, and uses that
-///  to set the const value for %7. In other words, this method propagates const
-///  info from result to operand (e.g. from %8 to %7), while
-///  getConstAddrAndLoadResult() propagates const info from operand to result.
+/// Otherwise, sets the memory at `addr` to an unspecified value and returns
+/// {false, unknown}, where `unknown` is an unknown SymbolicValue.
 ///
-///  As such, when const-evaluating an address-typed inst such as
-///  pointer_to_address, if the address is to be written to, caller should call
-///  this method (e.g. a[3] = 17). If the address is to be read (e.g. let v =
-///  a[3]), call getConstAddrAndLoadResult().
-SymbolicValue
-ConstExprFunctionState::getSingleWriterAddressValue(SILValue addr) {
-  // Check to see if we already have an answer.
-  auto it = calculatedValues.find(addr);
-  if (it != calculatedValues.end())
-    return it->second;
+/// Precondition: `memoryObject, accessPath` must be the constant value for
+/// `addr`.
+ConstExprFunctionState::SingleWriterAddressInitializationResult
+ConstExprFunctionState::initializeSingleWriterAddress(
+    SILValue addr, SymbolicValueMemoryObject *memoryObject,
+    ArrayRef<unsigned> accessPath) {
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr: initializeSingleWriterAddress " << addr);
+
+  // Keeps track of whether we actually initialized the memory.
+  bool initialized = false;
+
+  // If we detect instructions that initialize an aggregate piecewise, then we
+  // set this flag, which tells us to verify that the entire aggregate has been
+  // initialized.
+  bool mustCheckAggregateInitialized = false;
+
+  // Sets the pointed-at memory to `value`.
+  auto setMemoryValue = [&](SymbolicValue value) {
+    memoryObject->setIndexedElement(accessPath, value,
+                                    evaluator.getAllocator());
+  };
 
-  assert(addr->getType().isAddress());
-  auto *addrInst = dyn_cast<SingleValueInstruction>(addr);
-  if (!addrInst)
-    return evaluator.getUnknown(addr, UnknownReason::Default);
+  // Gets the pointed-at memory value.
+  auto getMemoryValue = [&]() -> SymbolicValue {
+    return memoryObject->getIndexedElement(accessPath);
+  };
 
-  // Keep track of the value found for the first constant store.
-  auto memoryAddress =
-      createMemoryObject(addr, SymbolicValue::getUninitMemory());
-  auto *memoryObject = memoryAddress.getAddressValueMemoryObject();
+  // Checks that the pointed-at aggregate is fully initialized.
+  // Precondition: The pointed-at memory value is uninit memory or an
+  // aggregate.
+  auto checkAggregateInitialized = [&]() -> bool {
```

Comment:
should we define this within the block of `if (auto *teai = dyn_cast<TupleElementAddrInst>(user))` below?

Reply:
No, it also gets used in the very outer block of the function, at the end.